scipy/cluster/hierarchy.py

```diff
@@ -3855,26 +3850,19 @@ def is_isomorphic(T1, T2):
     if T1.shape != T2.shape:
         raise ValueError('T1 and T2 must have the same number of elements.')
 
-    def py_is_isomorphic(T1, T2):
-        d1 = {}
-        d2 = {}
-        for t1, t2 in zip(T1, T2):
-            if t1 in d1:
-                if t2 not in d2:
-                    return False
-                if d1[t1] != t2 or d2[t2] != t1:
-                    return False
-            elif t2 in d2:
-                return False
-            else:
-                d1[t1] = t2
-                d2[t2] = t1
-        return True
+    # For each pair of (i, j) indices, test that
+    # T1[i] == T1[j] <--> T2[i] == T2[j]
+
+    # O(n*log(n)) algorithm.
+    # It is also possible to write a O(n) algorithm on top of unique_all(),
+    # but in practice it's much slower even for large clusters.
```

Comment:
tested up to 1m observations, 5m nodes. It just never seems to catch up.

Reply:
> As if `unique_all` was not implemented on top of a hashmap...

It is now:) https://numpy.org/devdocs/release/2.3.0-notes.html#performance-improvements-to-np-unique